README.md
# Riff Wave Reader

[![Build Status](https://travis-ci.org/lewismoten/riff-wave-reader.svg?branch=master)](https://travis-ci.org/lewismoten/riff-wave-reader)
[![npm version](https://badge.fury.io/js/riff-wave-reader.svg)](https://www.npmjs.com/package/riff-wave-reader)
[![install size](https://packagephobia.now.sh/badge?p=riff-wave-reader)](https://packagephobia.now.sh/result?p=riff-wave-reader)

This library reads the data within RIFF file with it's contents formatted as a WAVE file containing PCM data.

[Live Demo on GitHub ![Example Waveform](./docs/example-waveform.png)](https://lewismoten.github.io/riff-wave-reader/)

# Installation

This npm package is available as [riff-wave-reader](https://www.npmjs.com/package/riff-wave-reader)

From your terminal, install with the following command.

`npm install riff-wave-reader --save`

# How to use

## Node

```javascript
import Reader from "riff-wave-reader/reader";
let reader;
const channel = 0;
const index = 0;

// from array
const data =
  "52 49 46 46 d5 10 00 00 57 41 56 45 66 6d 74 20 " +
  "10 00 00 00 01 00 01 00 40 1f 00 00 40 1f 00 00 " +
  "01 00 08 00 64 61 74 61 b1 10 00 00 7f";
const array = data.split(" ").map(v => parseInt(v, 16));
reader = new RiffWaveReader(array);

// To handle large files without loading them completely into memory
// from wrapped file, buffer, array, and array buffers
reader = new RiffWaveReader(new Reader("./samples/hello.wav"));
reader = new RiffWaveReader(new Reader(Buffer.from(array)));
reader = new RiffWaveReader(new Reader(array));
reader = new RiffWaveReader(new ArrayBuffer(array));

// Read header chunks
reader.readChunks().then(chunks => {
  console.log(chunks);
});
reader.readSample(channel, index).then(sample => console.log(sample));
// 127
```

The chunks would be written out as:

```json
{
  "riff": {
    "tag": "RIFF",
    "size": 4309,
    "format": "WAVE"
  },
  "format": {
    "id": "fmt ",
    "size": 16,
    "type": 1,
    "channels": 1,
    "sampleRate": 8000,
    "byteRate": 8000,
    "blockAlignment": 1,
    "bitsPerSample": 8,
    "typeName": "PCM",
    "sampleSize": 1
  },
  "data": {
    "id": "data",
    "size": 4273,
    "start": 44,
    "sampleCount": 4265,
    "duration": 0.533125
  }
}
```

## Web Browser

```html
<script src="riff-wave-reader.js" type="text/javascript"></script>
<input type="file" change="changeFile(this.files)" />
<script type="text/javascript">
  function changeFile(files) {
    const fileReader = new FileReader();
    fileReader.onload = function(e) {
      var RiffWaveReader = window["riff-wave-reader"].RiffWaveReader;
      var riffReader = new RiffWaveReader(e.target.result);
      riffReader.readChunks().then(function(chunks) {
        console.log(chunks);
      });
    };
    fileReader.readAsArrayBuffer(files[0]);
  }
</script>
```

# Modifications

## Building

Building is done via webpack as a universal module definition (UMD) library.

From terminal

```bash
npm run build
```

1. Creates a script in development mode under /docs/dist/riff-wave-reader.js
1. Creates a script in production mode under /lib/dist/riff-wave-reader.js along with a sourcemap file

## Local Testing

From terminal

```bash
npm run start
```

Your web browser will open http://localhost:3000/docs/index.html


CHANGELOG.md
# Version 0

# 0.3

## 0.3.7: September 14, 2019
- GH-8 Support 24-bit

## 0.3.6: September 13, 2019
- GH-69 NPM Package / Security alert resolution

## 0.3.5: September 1, 2019

- GH-47 Document node `start` and `build` scripts
- Launch web browser to show demo via npm `start` script
- Resolve cross platform mapping on static local web server
- Save production source maps as a separate file

## 0.3.4: May 12, 2019

- Fix node support

## 0.3.3: May 12, 2019

- Fixed main entry file

## 0.3.2: May 12, 2019

- GH-2 Fixed signed/unsigned conversions
- GH-2 Support 16-bit
- GH-2 Use webpack to make browser-friendly
- GH-15 Test 8-bit 16khz
- GH-4 Get stereo working, read array buffer as unsigned bytes

## 0.3.1: May 9, 2019

- GH-38 Add package size badge
- GH-54 Remove package bloat
- GH-40 Doc npm install, add badge
- GH-40 Fix readme live demo link from GH-57
- GH-40 Fix readme formatting code block GH-57
- GH-40 Update web browser example due to GH-46
- GH-40 Fix changelog typo due to GH-54

## 0.3.0: May 7, 2019

- GH-57 Fix sample count calculation

# 0.2

## 0.2.2: May 6, 2019

- GH-54 Link to live demo
- GH-50 Show wave form in live demo
- GH-46 Include reader in live demo
- GH-44 Cache chunks

## 0.2.1: May 5, 2019

- GH-14 Load wave file via live demo
- GH-14 Read from ArrayBuffer

## 0.2.0: May 5, 2019

- GH-13 Read from array
- GH-42 Read large files/buffers/array from separate reader
- GH-14 Reduce dependencies in prep for browser environment
- GH-41 Read from riff/format/data chunks at once
- GH-21 Add doc on usage

# 0.1

## 0.1.1: May 4, 2019

- Support non-ES5 environment

## 0.1.0: May 4, 2019

- Initial release. The library can read an 8 bit wave files header chunks and individual samples.


src/index.js
import {
  ascii,
  uint8,
  uint16,
  int16,
  uint32,
  int24
} from "./converter";

const errorRiffTag = "RIFF chunk has wrong tag.";
const errorRiffFormat = "RIFF chunk specifies invalid format";
const errorFormatId = "Format chunk id is invalid";
const unknown = "Unknown";
const errorDataId = "Data chunk id is invalid";

export class RiffWaveReader {
  constructor(reader) {
    this.reader = reader;
  }

  _read(offset, size) {
    const reader = this.reader;
    if (Array.isArray(reader)) {
      return new Promise(resolve =>
        resolve(reader.slice(offset, offset + size + 1))
      );
    } else if (
      reader.constructor &&
      reader.constructor.name === "ArrayBuffer"
    ) {
      return new Promise(resolve =>
        resolve(new Uint8Array(reader.slice(offset, offset + size + 1)))
      );
    }
    return reader.read(offset, size);
  }

  readSample(channel, index) {
    return this.readChunks().then(({ format: {sampleSize, bitsPerSample}, data }) => {
      const position =
        data.start +
        index * sampleSize +
        (channel * bitsPerSample) / 8;
      const size = bitsPerSample / 8;
      return this._read(position, size).then(buffer => {
        switch(size) {
          default:
          case 1:return uint8(buffer, 0);
          case 2: return int16(buffer, 0);
          case 3: return int24(buffer, 0);
        }

      });
    });
  }
  readChunks() {
    if (this._chunks) {
      return new Promise(resolve => resolve(this._chunks));
    }
    return this._read(0, 44).then(buffer => {
      // RIFF
      const tag = ascii(buffer, 0, 4);
      if (tag !== "RIFF") throw errorRiffTag;

      let riffSize = uint32(buffer, 4);

      const format = ascii(buffer, 8, 4);
      if (format !== "WAVE") errorRiffFormat;

      const riffChunk = { tag, size: riffSize, format };

      // Format
      const id = ascii(buffer, 12, 4);
      if (id !== "fmt ") throw errorFormatId;
      const formatSize = uint32(buffer, 16);
      const type = uint16(buffer, 20);
      const channels = uint16(buffer, 22);
      const sampleRate = uint32(buffer, 24);
      const byteRate = uint32(buffer, 28);
      const blockAlignment = uint16(buffer, 32);
      const bitsPerSample = uint16(buffer, 34);

      // Calculations
      const typeName = type === 1 ? "PCM" : unknown;
      const sampleSize = (channels * bitsPerSample) / 8;

      const tagSize = 4;
      const lengthSize = 4;
      const tlvSize = tagSize + lengthSize;
      const riffChunkSize = tlvSize + 4; // WAVE
      const formatChunkSize = tlvSize + formatSize;
      const dataChunkStart = riffChunkSize + formatChunkSize;

      const formatChunk = {
        id,
        size: formatSize,
        type,
        channels,
        sampleRate,
        byteRate,
        blockAlignment,
        bitsPerSample,
        typeName,
        sampleSize
      };

      // Data
      let dataChunk;
      const dataId = ascii(buffer, dataChunkStart, tagSize);
      if (dataId !== "data") throw errorDataId;
      const dataSize = uint32(buffer, dataChunkStart + tagSize);
      const sampleStart = dataChunkStart + tlvSize;
      const sampleCount = dataSize / blockAlignment;
      const duration = sampleCount / sampleRate;
      dataChunk = {
        id: dataId,
        size: dataSize,
        start: sampleStart,
        sampleCount,
        duration
      };

      return (this._chunks = {
        riff: riffChunk,
        format: formatChunk,
        data: dataChunk
      });
    });
  }
}
export default RiffWaveReader;
